src/ipa/rpi/controller/rpi/clock_recovery.cpp

```diff
@@ -0,0 +1,140 @@
+#include "clock_recovery.h"
+
+#include <cctype>
+#include <chrono>
+#include <fcntl.h>
+#include <list>
+#include <map>
+#include <strings.h>
+#include <unistd.h>
+#include <vector>
+
+#include <libcamera/base/log.h>
+
+#include "sync_status.h"
+
+using namespace std::chrono_literals;
+using namespace RPiController;
+using namespace libcamera;
+
+LOG_DEFINE_CATEGORY(RPiClockRec)
+
+
+float ClockRecovery::slope()
+{
+	//returns slope for least squares line through points
+	double sl = (error_values.size() * xysum_ - xsum_ * ysum_) / (error_values.size() * x2sum_ - xsum_ * xsum_);
+	return sl;
+}
+
+void ClockRecovery::update_values(std::chrono::microseconds correction)
```

Comment:
Rather annoyingly, I think all these will have to be camel case too. So `ClockRecovery::updateValues`. Unfortunately this will apply to all functions... sorry :frowning_face: